# Cleaning the weather  data#

According to the documentation we have a daily weather table

In [448]:
import pandas as pd
import numpy as np

Load the weather data

In [449]:
weather = pd.read_csv('./weather.csv')

This is what the five first rows look like

In [450]:
weather.head()

Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1  2007-05-01    83    50   67     14        51      56    0    2   
1        2  2007-05-01    84    52   68      M        51      57    0    3   
2        1  2007-05-02    59    42   51     -3        42      47   14    0   
3        2  2007-05-02    60    43   52      M        42      47   13    0   
4        1  2007-05-03    66    46   56      2        40      48    9    0   

     ...    CodeSum Depth Water1 SnowFall PrecipTotal StnPressure SeaLevel  \
0    ...                0      M      0.0        0.00       29.10    29.82   
1    ...                M      M        M        0.00       29.18    29.82   
2    ...         BR     0      M      0.0        0.00       29.38    30.09   
3    ...      BR HZ     M      M        M        0.00       29.44    30.08   
4    ...                0      M      0.0        0.00       29.39    30.12   

  ResultSpeed ResultDir  AvgSpeed  
0         1.7        27       9.2  
1         2.7        25       9.6  
2        13.0         4      13.4  
3        13.3         2      13.4  
4        11.7         7      11.9  

[5 rows x 22 columns]

What columns do we have?

In [451]:
weather.columns

Index(['Station', 'Date', 'Tmax', 'Tmin', 'Tavg', 'Depart', 'DewPoint',
       'WetBulb', 'Heat', 'Cool', 'Sunrise', 'Sunset', 'CodeSum', 'Depth',
       'Water1', 'SnowFall', 'PrecipTotal', 'StnPressure', 'SeaLevel',
       'ResultSpeed', 'ResultDir', 'AvgSpeed'],
      dtype='object')

**Get more info about the weather data frame**

In [452]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2944 entries, 0 to 2943
Data columns (total 22 columns):
Station        2944 non-null int64
Date           2944 non-null object
Tmax           2944 non-null int64
Tmin           2944 non-null int64
Tavg           2944 non-null object
Depart         2944 non-null object
DewPoint       2944 non-null int64
WetBulb        2944 non-null object
Heat           2944 non-null object
Cool           2944 non-null object
Sunrise        2944 non-null object
Sunset         2944 non-null object
CodeSum        2944 non-null object
Depth          2944 non-null object
Water1         2944 non-null object
SnowFall       2944 non-null object
PrecipTotal    2944 non-null object
StnPressure    2944 non-null object
SeaLevel       2944 non-null object
ResultSpeed    2944 non-null float64
ResultDir      2944 non-null int64
AvgSpeed       2944 non-null object
dtypes: float64(1), int64(5), object(16)
memory usage: 506.1+ KB


**About the data**<BR />
2944 rows<BR />
22 columns<BR />
0 NULL values in the entire data set<BR />
Different data types in the columns<BR />
The columns need to be investigated closer.<BR />
The pdf that belongs to this data can be very usefull when we are going to use the data.

**Outlines for cleaning the data:**
- If more than 5% of row in a column are missing values -> drop the rows.
- If a column only consist of missing data, drop the column.
- For thre M/- that are left, convert into np.NaN (works as a float)
- Turn T into 0.1
- All numeric values should be a float (use round, if needed)

**Station** - OK

In [453]:
weather['Station'].dtype

dtype('int64')

The data type is ok and do not need to be changed.

In [454]:
weather['Station'].unique()

array([1, 2])

Seems like the data only comes from 2 different station.<BR />
This is good, this means that this column is binary.<BR />
This column is no described in the documentation. I can also be that 1 and 2 stand for:<BR />
AO1; automated station without a precipitation descriminator.<BR />
AO2; automated station with precipitation descriminator.

Turn into floats

In [455]:
weather['Station'] = weather['Station'].astype(float)
weather['Station'].head()

0    1.0
1    2.0
2    1.0
3    2.0
4    1.0
Name: Station, dtype: float64

**Date** - OK

In [456]:
weather['Date'].dtype

dtype('O')

The data type is an object.<BR />
We want to turn it into datetime-format instead.

In [457]:
# Look at the format of the date, looks like YYYY-mm-dd
weather['Date'].unique()

array(['2007-05-01', '2007-05-02', '2007-05-03', ..., '2014-10-29',
       '2014-10-30', '2014-10-31'], dtype=object)

In [458]:
# Set a big Y so that pandas know that the the year is 4 and not 2 digits 
weather['Date'] = pd.to_datetime(weather['Date'], format='%Y-%m-%d')

In [459]:
# Check and see that the data type changed
weather['Date'].unique()

array(['2007-05-01T00:00:00.000000000', '2007-05-02T00:00:00.000000000',
       '2007-05-03T00:00:00.000000000', ...,
       '2014-10-29T00:00:00.000000000', '2014-10-30T00:00:00.000000000',
       '2014-10-31T00:00:00.000000000'], dtype='datetime64[ns]')

Make new columns in data frame for Year, Month, Day

In [460]:
weather['Year'] = weather['Date'].dt.year

In [461]:
weather['Month'] = weather['Date'].dt.month

In [462]:
weather['Day'] = weather['Date'].dt.day

In [463]:
# The new columns were added as the last columns in the data frame
# All 3 columns are the data type int64

Over what time period is the data collected?

In [464]:
# Earliest collection
print(weather['Date'].min())
# Latest collection
print(weather['Date'].max())

2007-05-01 00:00:00
2014-10-31 00:00:00


Comment: datetime can not be converted into floats

**Tmax** - OK<BR />
Maximum degrees Fahrenheit

In [465]:
weather['Tmax'].dtype

dtype('int64')

The data type is ok and do not need to be changed.

In [466]:
np.sort(weather['Tmax'].unique())

array([ 41,  42,  44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,
        55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,
        68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,
        81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,  93,
        94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104])

Takes on values from 41 to 104 F.

No NULL values, M or -

Turn into floats

In [467]:
weather['Tmax'] = weather['Tmax'].astype(float)
weather['Tmax'].head()

0    83.0
1    84.0
2    59.0
3    60.0
4    66.0
Name: Tmax, dtype: float64

**Tmin** - OK<BR />
Minimum degrees Fahrenheit

In [468]:
weather['Tmin'].dtype

dtype('int64')

The data type is ok and do not need to be changed.

In [469]:
np.sort(weather['Tmin'].unique())

array([29, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46,
       47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63,
       64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80,
       81, 82, 83])

Takes on values from 29 to 83 F.

No NULL values, M or -

Turn into floats

In [470]:
weather['Tmin'] = weather['Tmin'].astype(float)
weather['Tmin'].head()

0    50.0
1    52.0
2    42.0
3    43.0
4    46.0
Name: Tmin, dtype: float64

**Tavg** - OK<BR />
Average degrees Fahrenheit

In [471]:
weather['Tavg'].dtype

dtype('O')

The data type is object, we need to find out why. We want it to be numeric.

In [472]:
np.sort(weather['Tavg'].unique())

array(['36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46',
       '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57',
       '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68',
       '69', '70', '71', '72', '73', '74', '75', '76', '77', '78', '79',
       '80', '81', '82', '83', '84', '85', '86', '87', '88', '89', '90',
       '91', '92', '93', '94', 'M'], dtype=object)

We hava a M instead of a number in our data.<BR />
According to the documentation M = Missing Data, or data that are not normally reported for the station.

In [473]:
weather['Tavg'].value_counts()

73    138
77    117
70    117
75    110
71    109
74    107
72    104
69    103
78    102
76    100
68     99
79     98
66     93
67     89
61     88
64     86
65     84
80     84
63     81
57     67
62     66
60     61
50     57
81     55
53     49
58     49
82     48
54     48
55     48
56     46
52     46
59     45
51     36
83     34
49     29
45     28
46     24
47     24
84     21
44     19
48     17
85     16
86     16
42     15
43     12
M      11
87      9
41      7
40      5
88      4
39      4
91      4
89      4
36      2
37      2
38      2
90      2
93      1
92      1
94      1
Name: Tavg, dtype: int64

There are 11 rows out of 2944 rows with missing values for the average temperature.

Drop the rows with missing data (by making new dataframe only containing the rows that were not M for Tavg)

In [474]:
weather = weather[weather['Tavg'] != 'M']

Turn into floats

In [475]:
weather['Tavg'] = weather['Tavg'].astype(float)
weather['Tavg'].head()

0    67.0
1    68.0
2    51.0
3    52.0
4    56.0
Name: Tavg, dtype: float64

**Depart** - OK<BR />
This column indicates DEPARTURE FROM NORMAL

In [476]:
weather['Depart'].dtype

dtype('O')

The data type is object, we need to find out why. We want it to be numeric.

In [477]:
np.sort(weather['Depart'].unique())

array([' 0', ' 1', ' 2', ' 3', ' 4', ' 5', ' 6', ' 7', ' 8', ' 9', '-1',
       '-10', '-11', '-12', '-13', '-14', '-15', '-16', '-17', '-2', '-3',
       '-4', '-5', '-6', '-7', '-8', '-9', '10', '11', '12', '13', '14',
       '15', '16', '17', '18', '19', '20', '21', '22', '23', 'M'],
      dtype=object)

We hava a M instead of a number in our data.<BR />
According to the documentation M = Missing Data, or data that are not normally reported for the station.

In [478]:
weather['Depart'].value_counts()

M      1461
 2       93
-1       84
-2       80
 5       77
 1       76
 7       76
 3       75
 0       74
-3       72
 4       71
 6       67
 8       59
-5       57
-4       56
-6       50
 9       47
10       46
-8       43
-7       30
12       28
11       28
-9       25
13       23
-10      22
14       22
15       15
16       12
-11      10
-12       8
17        7
-14       6
18        6
-13       5
20        4
19        4
22        3
-15       3
-16       3
-17       2
21        2
23        1
Name: Depart, dtype: int64

There are 1472 rows out of 2944 rows with missing values for the depart column.

Replace all the M with np.nan

In [479]:
weather['Depart'] = weather['Depart'].replace(['M'], np.nan)

Turn into floats

In [480]:
weather['Depart'] = weather['Depart'].astype(float)
weather['Depart'].head()

0    14.0
1     NaN
2    -3.0
3     NaN
4     2.0
Name: Depart, dtype: float64

**DewPoint**<BR />
This column give us the AVERAGE DEW POINT

In [481]:
weather['DewPoint'].dtype

dtype('int64')

The data type is ok and do not need to be changed.

In [482]:
np.sort(weather['DewPoint'].unique())

array([22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
       39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55,
       56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72,
       73, 74, 75])

Takes on values from 22 to 75 F.

No NULL values, M or -

Turn into floats

In [483]:
weather['DewPoint'] = weather['DewPoint'].astype(float)
weather['DewPoint'].head()

0    51.0
1    51.0
2    42.0
3    42.0
4    40.0
Name: DewPoint, dtype: float64

**WetBulb** - OK<BR />
This column give us the AVERAGE WET BULB<BR />
Explenation to Wet-Bulb here https://en.wikipedia.org/wiki/Wet-bulb_temperature

In [484]:
weather['WetBulb'].dtype

dtype('O')

The data type is object, we need to find out why. We want it to be numeric.

In [485]:
np.sort(weather['WetBulb'].unique())

array(['32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42',
       '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53',
       '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64',
       '65', '66', '67', '68', '69', '70', '71', '72', '73', '74', '75',
       '76', '77', '78', 'M'], dtype=object)

We hava a M instead of a number in our data.<BR />
According to the documentation M = Missing Data, or data that are not normally reported for the station.

In [486]:
weather['WetBulb'].value_counts()

63    135
65    131
59    129
61    122
64    121
62    118
67    117
66    113
60    111
69    107
57    105
70    102
55     99
68     98
71     97
58     95
56     95
52     78
46     72
72     71
53     69
47     59
51     57
73     56
54     54
50     53
45     52
48     51
49     50
74     49
44     45
42     35
43     31
40     31
75     20
41     19
38     17
39     15
36      9
77      8
76      8
35      8
33      5
37      5
34      5
M       4
32      1
78      1
Name: WetBulb, dtype: int64

There are 4 rows out of 2944 rows with missing values for the WetBulb column.

Drop the rows with missing data (by making new dataframe only containing the rows that were not M for WetBulb)

In [487]:
weather = weather[weather['WetBulb'] != 'M']

Turn into floats

In [488]:
weather['WetBulb'] = weather['WetBulb'].astype(float)
weather['WetBulb'].head()

0    56.0
1    57.0
2    47.0
3    47.0
4    48.0
Name: WetBulb, dtype: float64

In [489]:
# ignoring the error message since the data type was successfully changed.
weather.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2929 entries, 0 to 2943
Data columns (total 25 columns):
Station        2929 non-null float64
Date           2929 non-null datetime64[ns]
Tmax           2929 non-null float64
Tmin           2929 non-null float64
Tavg           2929 non-null float64
Depart         1469 non-null float64
DewPoint       2929 non-null float64
WetBulb        2929 non-null float64
Heat           2929 non-null object
Cool           2929 non-null object
Sunrise        2929 non-null object
Sunset         2929 non-null object
CodeSum        2929 non-null object
Depth          2929 non-null object
Water1         2929 non-null object
SnowFall       2929 non-null object
PrecipTotal    2929 non-null object
StnPressure    2929 non-null object
SeaLevel       2929 non-null object
ResultSpeed    2929 non-null float64
ResultDir      2929 non-null int64
AvgSpeed       2929 non-null object
Year           2929 non-null int64
Month          2929 non-null int64
Day            2

**Heat** - OK<BR />
According to the documentation:<BR />
Degree Days: Base 65 F<BR />
HEATING (SEASON BEGINS WITH JULY)

In [490]:
weather['Heat'].dtype

dtype('O')

The data type is object, we need to find out why. We want it to be numeric.

In [491]:
np.sort(weather['Heat'].unique())

array(['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18',
       '19', '2', '20', '21', '22', '23', '24', '25', '26', '27', '28',
       '29', '3', '4', '5', '6', '7', '8', '9'], dtype=object)

We hava a M instead of a number in our data.<BR />
According to the documentation M = Missing Data, or data that are not normally reported for the station.

In [492]:
weather['Heat'].value_counts()

0     1866
4       88
1       86
2       81
8       67
3       66
5       61
15      57
12      49
7       49
10      48
11      48
9       46
13      46
6       45
14      36
16      29
20      28
18      24
19      24
21      19
17      17
23      15
22      12
24       7
25       5
26       4
27       2
28       2
29       2
Name: Heat, dtype: int64

There are 11 rows out of 2944 rows with missing values for the Heat column.

Drop the rows with missing data (by making new dataframe only containing the rows that were not M for Heat)

In [493]:
weather = weather[weather['Heat'] != 'M']

Turn into floats

In [494]:
weather['Heat'] = weather['Heat'].astype(float)
weather['Heat'].head()

0     0.0
1     0.0
2    14.0
3    13.0
4     9.0
Name: Heat, dtype: float64

**Cool** - OK<BR />
According to the documentation:<BR />
Degree Days: Base 65 F<BR />
COOLING (SEASON BEGINS WITH JANUARY)

In [495]:
weather['Cool'].dtype

dtype('O')

The data type is object, we need to find out why. We want it to be numeric.

In [496]:
np.sort(weather['Cool'].unique())

array([' 0', ' 1', ' 2', ' 3', ' 4', ' 5', ' 6', ' 7', ' 8', ' 9', '10',
       '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21',
       '22', '23', '24', '25', '26', '27', '28', '29'], dtype=object)

We hava a M instead of a number in our data.<BR />
According to the documentation M = Missing Data, or data that are not normally reported for the station.

In [497]:
weather['Cool'].value_counts()

 0    1147
 8     137
 5     117
12     116
10     110
 6     108
 9     107
 7     104
 4     103
13     101
11     100
 3      99
14      98
 1      93
 2      89
15      84
16      55
17      48
18      34
19      21
20      16
21      16
22       9
26       4
24       4
23       4
25       2
29       1
28       1
27       1
Name: Cool, dtype: int64

There are 11 rows out of 2944 rows with missing values for the Heat column.

Drop the rows with missing data (by making new dataframe only containing the rows that were not M for Cool)

In [498]:
weather = weather[weather['Cool'] != 'M']

Turn into floats

In [499]:
weather['Cool'] = weather['Cool'].astype(float)
weather['Cool'].head()

0    2.0
1    3.0
2    0.0
3    0.0
4    0.0
Name: Cool, dtype: float64

**Sunrise** - OK<BR />
According to the documentation:<BR />
SUNRISE  (Calculated, not observed)

In [500]:
weather['Sunrise'].dtype

dtype('O')

The data type is object, we need to find out why. We want it to be numeric.

In [501]:
np.sort(weather['Sunrise'].unique())

array(['-', '0416', '0417', '0418', '0419', '0420', '0421', '0422',
       '0423', '0424', '0425', '0426', '0427', '0428', '0429', '0430',
       '0431', '0432', '0433', '0434', '0435', '0436', '0437', '0438',
       '0439', '0440', '0441', '0442', '0443', '0444', '0445', '0446',
       '0447', '0448', '0449', '0450', '0451', '0452', '0453', '0454',
       '0455', '0456', '0457', '0458', '0459', '0500', '0501', '0502',
       '0503', '0504', '0505', '0506', '0508', '0509', '0510', '0511',
       '0512', '0513', '0514', '0515', '0516', '0517', '0518', '0519',
       '0520', '0521', '0522', '0523', '0524', '0525', '0526', '0527',
       '0528', '0529', '0530', '0531', '0532', '0533', '0535', '0536',
       '0537', '0538', '0539', '0540', '0541', '0542', '0543', '0544',
       '0545', '0546', '0547', '0548', '0549', '0550', '0552', '0553',
       '0554', '0555', '0556', '0557', '0558', '0559', '0600', '0602',
       '0603', '0604', '0605', '0606', '0607', '0608', '0610', '0611',
       '0

We hava a - instead of a number in our data.<BR />
According to the documentation - = values not available for sunrise/sunset.

In [502]:
weather['Sunrise'].value_counts()

-       1460
0416     104
0417      64
0419      40
0420      32
0422      32
0425      32
0418      31
0433      24
0428      24
0423      24
0421      24
0430      16
0436      16
0443      16
0429      16
0442      16
0437      16
0432      16
0424      16
0441      16
0439      16
0444      16
0427      16
0448      16
0447      16
0434      16
0438      16
0431      16
0426      16
        ... 
0550       8
0612       8
0614       8
0453       8
0513       8
0610       8
0529       8
0600       8
0506       8
0554       8
0611       8
0548       8
0536       8
0616       8
0545       8
0618       8
0559       8
0608       8
0539       8
0538       8
0622       8
0604       8
0525       8
0549       8
0557       8
0528       8
0620       8
0535       8
0455       7
0454       7
Name: Sunrise, Length: 122, dtype: int64

There are 1472 rows out of 2944 rows with missing values for the Sunrise column.
#### Keep the column but do not do anything with it. ####

In [503]:
# I case we want to use it, we are going to handle the format and the -

**Sunset** - OK<BR />
According to the documentation:<BR />
SUNSET   (Calculated, not observed)

In [504]:
weather['Sunset'].dtype

dtype('O')

The data type is object, we need to find out why. We want it to be numeric.

In [505]:
np.sort(weather['Sunset'].unique())

array(['-', '1647', '1649', '1650', '1651', '1653', '1654', '1656',
       '1657', '1658', '1660', '1701', '1703', '1704', '1706', '1707',
       '1709', '1711', '1712', '1714', '1715', '1717', '1719', '1720',
       '1722', '1724', '1725', '1727', '1729', '1730', '1732', '1734',
       '1735', '1737', '1739', '1741', '1742', '1744', '1746', '1748',
       '1749', '1751', '1753', '1755', '1756', '1758', '1760', '1803',
       '1805', '1807', '1809', '1810', '1812', '1814', '1815', '1817',
       '1819', '1821', '1822', '1824', '1826', '1827', '1829', '1831',
       '1832', '1834', '1835', '1837', '1839', '1840', '1842', '1843',
       '1845', '1846', '1848', '1849', '1850', '1851', '1852', '1853',
       '1854', '1855', '1856', '1857', '1858', '1859', '1860', '1900',
       '1901', '1902', '1903', '1904', '1905', '1906', '1907', '1908',
       '1909', '1910', '1911', '1912', '1913', '1914', '1915', '1916',
       '1917', '1918', '1919', '1920', '1921', '1922', '1923', '1924',
       '1

We hava a - instead of a number in our data.<BR />
According to the documentation - = values not available for sunrise/sunset.

In [506]:
weather['Sunset'].value_counts()

-       1460
1931      95
1930      56
1929      48
1923      32
1925      32
1927      32
1928      32
1918      24
1916      24
1926      24
1920      24
1907      16
1902      16
1919      16
1917      16
1924      16
1909      16
1921      16
1915      16
1914      16
1904      16
1910      16
1903      16
1855      16
1922      16
1906      16
1851      16
1858      16
1856      16
        ... 
1741       8
1901       8
1657       8
1654       8
1809       8
1751       8
1709       8
1822       8
1758       8
1739       8
1744       8
1846       8
1819       8
1803       8
1651       8
1912       8
1850       8
1756       8
1810       8
1717       8
1817       8
1712       8
1714       8
1660       8
1857       8
1658       8
1834       8
1755       8
1753       8
1900       7
Name: Sunset, Length: 119, dtype: int64

There are 1472 rows out of 2944 rows with missing values for the Sunset column.
#### Keep the column but do not do anything with it. ####

In [507]:
# I case we want to use it, we are going to handle the format and the -

**CodeSum**<BR />
Significant Weather Types (Weather Phenomena)

Get unique values

In [508]:
np.sort(weather['CodeSum'].unique())

array([' ', 'BCFG BR', 'BR', 'BR HZ', 'BR HZ FU', 'BR HZ VCFG', 'BR VCTS',
       'DZ', 'DZ BR', 'DZ BR HZ', 'FG BR HZ', 'FG+', 'FG+ BCFG BR',
       'FG+ BR', 'FG+ BR HZ', 'FG+ FG BR', 'FG+ FG BR HZ', 'FG+ MIFG BR',
       'FU', 'HZ', 'HZ FU', 'HZ VCTS', 'MIFG BCFG BR', 'RA', 'RA BCFG BR',
       'RA BR', 'RA BR FU', 'RA BR HZ', 'RA BR HZ FU', 'RA BR HZ VCFG',
       'RA BR HZ VCTS', 'RA BR SQ', 'RA BR VCFG', 'RA BR VCTS', 'RA DZ',
       'RA DZ BR', 'RA DZ BR HZ', 'RA DZ FG+ BCFG BR', 'RA DZ FG+ BR',
       'RA DZ FG+ BR HZ', 'RA DZ FG+ FG BR', 'RA DZ SN', 'RA FG BR',
       'RA FG+ BR', 'RA FG+ MIFG BR', 'RA HZ', 'RA SN', 'RA SN BR',
       'RA VCTS', 'TS', 'TS BR', 'TS BR HZ', 'TS HZ', 'TS RA', 'TS RA BR',
       'TS RA BR HZ', 'TS RA FG+ FG BR', 'TS TSRA', 'TS TSRA BR',
       'TS TSRA BR HZ', 'TS TSRA GR RA BR', 'TS TSRA HZ', 'TS TSRA RA',
       'TS TSRA RA BR', 'TS TSRA RA BR HZ', 'TS TSRA RA BR HZ VCTS',
       'TS TSRA RA BR VCTS', 'TS TSRA RA FG BR', 'TS TSRA RA FG BR HZ',
 

By hand turn it into dummies

In [509]:
col_list = ['','BCFG','BR','DZ','FG','FG+','FU','GR','HZ','MIFG','RA','SN','SQ','TS','TSRA','VCFG','VCTS'] #unique values
df_temp = pd.DataFrame(columns = col_list) # take unique values and put as headers in new df
weather = pd.concat([weather, df_temp], axis = 1) # concat weather and new df

weather[col_list] = 0 # set all the new column to 0

In [510]:
for i, _ in weather.iterrows(): # use iterrows to get correct index (did not work with len(weather) bc index were missing row numbers since we removed rows)
    codesum_split = weather['CodeSum'][i].split(' ') # look at codesum and split at ' '
    for item in codesum_split:
        weather.set_value(i, item, 1) # put 1 instead of 0 in the right column if the string of letters existed.

In [511]:
weather.head()

Station       Date  Tmax  Tmin  Tavg  Depart  DewPoint  WetBulb  Heat  \
0      1.0 2007-05-01  83.0  50.0  67.0    14.0      51.0     56.0   0.0   
1      2.0 2007-05-01  84.0  52.0  68.0     NaN      51.0     57.0   0.0   
2      1.0 2007-05-02  59.0  42.0  51.0    -3.0      42.0     47.0  14.0   
3      2.0 2007-05-02  60.0  43.0  52.0     NaN      42.0     47.0  13.0   
4      1.0 2007-05-03  66.0  46.0  56.0     2.0      40.0     48.0   9.0   

   Cool  ...  GR HZ MIFG RA SN SQ TS TSRA VCFG  VCTS  
0   2.0  ...   0  0    0  0  0  0  0    0    0     0  
1   3.0  ...   0  0    0  0  0  0  0    0    0     0  
2   0.0  ...   0  0    0  0  0  0  0    0    0     0  
3   0.0  ...   0  1    0  0  0  0  0    0    0     0  
4   0.0  ...   0  0    0  0  0  0  0    0    0     0  

[5 rows x 42 columns]

**Depth** - OK<BR />
INCHES<BR />
I can not understand from the documentation what this is.<BR />
But I guess depth of snow/ice or rain maybe...

In [512]:
weather['Depth'].unique()

array(['0', 'M'], dtype=object)

We only have the data 0 inches and missing data. 

Drop this column from dataframe!

In [513]:
weather = weather.drop(['Depth'], axis=1)

# make sure the column is gone
weather.columns

Index(['Station', 'Date', 'Tmax', 'Tmin', 'Tavg', 'Depart', 'DewPoint',
       'WetBulb', 'Heat', 'Cool', 'Sunrise', 'Sunset', 'CodeSum', 'Water1',
       'SnowFall', 'PrecipTotal', 'StnPressure', 'SeaLevel', 'ResultSpeed',
       'ResultDir', 'AvgSpeed', 'Year', 'Month', 'Day', '', 'BCFG', 'BR', 'DZ',
       'FG', 'FG+', 'FU', 'GR', 'HZ', 'MIFG', 'RA', 'SN', 'SQ', 'TS', 'TSRA',
       'VCFG', 'VCTS'],
      dtype='object')

**Water1** - OK<BR />
WATER EQUIVALENT
I can not understand from the documentation what this is.

In [514]:
weather['Water1'].unique()

array(['M'], dtype=object)

This column only contains missing data. 

Drop this column from dataframe!

In [515]:
weather = weather.drop(['Water1'], axis=1)

# make sure the column is gone
weather.columns

Index(['Station', 'Date', 'Tmax', 'Tmin', 'Tavg', 'Depart', 'DewPoint',
       'WetBulb', 'Heat', 'Cool', 'Sunrise', 'Sunset', 'CodeSum', 'SnowFall',
       'PrecipTotal', 'StnPressure', 'SeaLevel', 'ResultSpeed', 'ResultDir',
       'AvgSpeed', 'Year', 'Month', 'Day', '', 'BCFG', 'BR', 'DZ', 'FG', 'FG+',
       'FU', 'GR', 'HZ', 'MIFG', 'RA', 'SN', 'SQ', 'TS', 'TSRA', 'VCFG',
       'VCTS'],
      dtype='object')

**SnowFall** - OK<BR />
INCHES (24-HR PERIOD ENDING AT INDICATED LOCAL STANDARD TIME)<BR />
SNOWFALL (INCHES AND TENTHS)(2400 LST)- Not all stations report snow totals<BR />
T = TRACE<BR />
M = MISSING DATA

In [516]:
weather['SnowFall'].unique()

array(['0.0', 'M', '  T', '0.1'], dtype=object)

In [517]:
weather['SnowFall'].value_counts()

M      1460
0.0    1456
  T      12
0.1       1
Name: SnowFall, dtype: int64

There are 1472 rows out of 2944 rows with missing values for the SnowFall column.

Replace '  T' with 0.1 and 'M' with np.nan

In [518]:
weather['SnowFall'] = weather['SnowFall'].replace(['  T'], '0.1')

In [519]:
weather['SnowFall'] = weather['SnowFall'].replace(['M'], np.nan)

In [520]:
weather['SnowFall'].unique()

array(['0.0', nan, '0.1'], dtype=object)

Turn into floats

In [521]:
weather['SnowFall'] = weather['SnowFall'].astype(float)
weather['SnowFall'].head()

0    0.0
1    NaN
2    0.0
3    NaN
4    0.0
Name: SnowFall, dtype: float64

**PrecipTotal** - OK<BR />
INCHES (24-HR PERIOD ENDING AT INDICATED LOCAL STANDARD TIME)<BR />
SNOWFALL (INCHES AND TENTHS)(2400 LST)- Not all stations report snow totals<BR />
AND<BR />
WATER EQUIVALENT(INCHES & HUNDREDTHS(2400 LST) RAINFALL & MELTED SNOW<BR />
T = TRACE<BR />
M = MISSING DATA

In [522]:
weather['PrecipTotal'].unique()

array(['0.00', '  T', '0.13', '0.02', '0.38', '0.60', '0.14', '0.07',
       '0.11', '0.09', '1.01', '0.28', '0.04', '0.08', '0.01', '0.53',
       '0.19', '0.21', '0.32', '0.39', '0.31', '0.42', '0.27', '0.16',
       '0.58', '0.93', '0.05', '0.34', '0.15', '0.35', 'M', '0.40',
       '0.66', '0.30', '0.24', '0.43', '1.55', '0.92', '0.89', '0.17',
       '0.03', '1.43', '0.97', '0.26', '1.31', '0.06', '0.46', '0.29',
       '0.23', '0.41', '0.45', '0.83', '1.33', '0.91', '0.48', '0.37',
       '0.88', '2.35', '1.96', '0.20', '0.25', '0.18', '0.67', '0.36',
       '0.33', '1.28', '0.74', '0.76', '0.71', '0.95', '1.46', '0.12',
       '0.52', '0.64', '0.22', '1.24', '0.72', '0.73', '0.65', '1.61',
       '1.22', '0.50', '1.05', '2.43', '0.59', '2.90', '2.68', '1.23',
       '0.62', '6.64', '3.07', '1.44', '1.75', '0.82', '0.80', '0.86',
       '0.63', '0.55', '1.03', '0.70', '1.73', '1.38', '0.44', '1.14',
       '1.07', '3.97', '0.87', '0.78', '1.12', '0.68', '0.10', '0.61',
       '0.

In [523]:
weather['PrecipTotal'].value_counts()

0.00    1570
  T      317
0.01     126
0.02      63
0.03      46
0.04      35
0.05      32
0.12      28
0.08      28
0.06      26
0.07      23
0.16      21
0.09      21
0.11      20
0.14      19
0.17      17
0.13      14
0.18      14
0.19      14
0.28      14
0.15      13
0.20      13
0.25      11
0.26      11
0.23      11
0.10      10
0.24      10
0.40       9
0.43       9
0.21       9
        ... 
3.15       1
1.30       1
2.09       1
2.43       1
3.07       1
2.24       1
0.90       1
3.97       1
1.07       1
3.17       1
0.96       1
1.82       1
2.90       1
6.86       1
3.64       1
0.91       1
2.76       1
1.34       1
6.64       1
1.24       1
1.19       1
1.21       1
1.18       1
1.57       1
1.60       1
2.06       1
1.11       1
1.96       1
2.35       1
2.60       1
Name: PrecipTotal, Length: 168, dtype: int64

We have to handle 318 T and some smaller number of M.
Since T stands for trace, that can be replaces by a small number.

Replace '  T' with 0.1 and 'M' with np.nan

In [524]:
weather['PrecipTotal'] = weather['PrecipTotal'].replace(['  T'], '0.1')

In [525]:
weather['PrecipTotal'] = weather['PrecipTotal'].replace(['M'], np.nan)

Turn into floats

In [526]:
weather['PrecipTotal'] = weather['PrecipTotal'].astype(float)
weather['PrecipTotal'].head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: PrecipTotal, dtype: float64

**StnPressure** - OK<BR />
Pressure, INCHES OF HG<BR />
AVERAGE STATION PRESSURE<BR />

In [527]:
weather['StnPressure'].unique()

array(['29.10', '29.18', '29.38', '29.44', '29.39', '29.46', '29.31',
       '29.40', '29.57', '29.62', '29.29', '29.36', '29.21', '29.28',
       '29.20', '29.26', '29.33', '29.49', '29.54', '29.55', '29.23',
       '29.13', '29.19', '29.53', '29.60', '29.34', '29.41', '29.47',
       '29.51', '29.42', '29.43', '29.25', '29.03', '28.82', '28.87',
       '28.88', '29.16', '29.07', '28.84', '28.91', '29.24', 'M', '29.30',
       '29.12', '29.45', '29.56', '29.32', '29.05', '29.11', '29.06',
       '29.22', '29.08', '29.14', '29.37', '29.35', '29.15', '29.17',
       '29.48', '29.52', '29.27', '29.50', '28.59', '28.67', '28.75',
       '29.02', '29.79', '29.86', '29.63', '29.70', '28.95', '29.01',
       '28.79', '28.85', '28.97', '28.89', '28.94', '28.93', '28.98',
       '28.96', '29.00', '29.66', '29.09', '28.90', '29.04', '29.59',
       '29.65', '29.58', '29.61', '29.64', '29.71', '29.67', '28.80',
       '28.73', '29.68', '28.74', '28.55', '28.63', '28.92', '28.99',
       '28.81',

In [528]:
weather['StnPressure'].value_counts()

29.34    128
29.28    124
29.26    123
29.21    107
29.31    106
29.23    103
29.36     94
29.41     90
29.39     88
29.18     86
29.29     86
29.24     82
29.33     80
29.44     70
29.38     64
29.13     64
29.15     62
29.20     60
29.47     57
29.16     54
29.25     52
29.10     47
29.35     47
29.49     46
29.42     41
29.08     41
29.19     39
29.40     37
29.32     36
29.30     36
        ... 
28.91      4
29.67      4
29.61      4
28.79      4
28.80      3
28.94      3
28.82      3
28.99      3
28.96      3
29.68      3
28.93      3
28.84      2
M          2
28.85      2
28.71      2
28.81      2
28.83      1
28.55      1
28.59      1
29.79      1
28.77      1
28.63      1
28.74      1
28.75      1
29.86      1
28.69      1
28.67      1
28.73      1
28.88      1
29.71      1
Name: StnPressure, Length: 104, dtype: int64

Only 4 missing values.

Drop the rows with missing data (by making new dataframe only containing the rows that were not M for StnPressure)

In [529]:
weather = weather[weather['StnPressure'] != 'M']

Turn into floats

In [530]:
weather['StnPressure'] = weather['StnPressure'].astype(float)
weather['StnPressure'].head()

0    29.10
1    29.18
2    29.38
3    29.44
4    29.39
Name: StnPressure, dtype: float64

**SeaLevel** - OK<BR />
Pressure, INCHES OF HG<BR />
AVERAGE SEA LEVEL PRESSURE<BR />

In [531]:
weather['SeaLevel'].unique()

array(['29.82', '30.09', '30.08', '30.12', '30.05', '30.10', '30.29',
       '30.28', '30.03', '30.02', '29.94', '29.93', '29.92', '29.91',
       '30.04', '30.20', '30.19', '30.24', '29.97', '29.98', '29.84',
       '29.83', '30.27', '30.25', '30.26', '30.11', '30.06', '30.23',
       '30.15', '30.14', '30.00', '29.99', '29.90', '29.77', '29.76',
       '29.56', '29.54', '29.52', '29.51', '29.79', '29.78', '29.81',
       '29.55', '29.85', '30.07', '30.16', '29.96', '29.95', '30.13',
       '30.21', '30.22', '29.88', '30.01', '29.80', '29.89', '29.74',
       '29.87', '29.86', '30.18', '30.17', '29.34', '29.44', '29.45',
       '29.71', '29.72', '30.52', '30.53', '30.40', '30.41', '29.67',
       '29.53', '29.69', '29.61', '29.64', '29.63', '29.66', '29.70',
       '30.34', '30.33', '29.62', '29.60', '29.75', '29.68', '29.73',
       '30.31', '30.30', '30.32', '30.37', '30.39', 'M', '29.59', '29.65',
       '30.35', '30.36', '29.48', '30.38', '29.50', '29.25', '29.23',
       '29.46',

In [532]:
weather['SeaLevel'].value_counts()

30.00    95
29.94    85
29.98    85
29.89    82
29.92    82
30.05    81
30.02    80
29.95    80
29.91    80
29.93    79
29.96    76
29.97    73
29.99    73
30.03    72
30.01    71
30.04    70
30.06    69
29.87    62
29.90    60
29.86    58
29.85    56
29.88    56
30.08    54
30.10    53
29.83    53
30.11    52
30.09    50
30.07    50
29.84    47
29.82    44
         ..
29.59     7
29.62     6
29.55     6
30.31     6
30.37     4
30.36     4
29.63     4
29.53     4
29.45     4
29.64     3
29.56     3
30.34     3
29.54     3
29.50     2
29.51     2
29.48     2
30.38     2
29.34     2
30.40     2
29.52     2
29.47     1
29.23     1
30.52     1
29.43     1
29.25     1
29.46     1
30.41     1
30.39     1
29.44     1
30.53     1
Name: SeaLevel, Length: 102, dtype: int64

Btw 7 and 44 missing values.

Drop the rows with missing data (by making new dataframe only containing the rows that were not M for SeaLevel)

In [533]:
weather = weather[weather['SeaLevel'] != 'M']

Turn into floats

In [534]:
weather['SeaLevel'] = weather['SeaLevel'].astype(float)
weather['SeaLevel'].head()

0    29.82
1    29.82
2    30.09
3    30.08
4    30.12
Name: SeaLevel, dtype: float64

**ResultSpeed** - OK<BR />
RESULTANT WIND SPEED

In [535]:
weather['ResultSpeed'].dtype

dtype('float64')

The data type is ok and do not need to be changed.

In [536]:
np.sort(weather['ResultSpeed'].unique())

array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
        7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
        8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
       10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
       11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
       12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
       13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
       14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15. , 15

**ResultDir** - OK<BR />
RESULTANT DIRECTION - (WHOLE DEGREES)

In [537]:
weather['ResultDir'].dtype

dtype('int64')

The data type need to be changed into floats

In [538]:
np.sort(weather['ResultDir'].unique())

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36])

Turn into floats

In [539]:
weather['ResultDir'] = weather['ResultDir'].astype(float)
weather['ResultDir'].head()

0    27.0
1    25.0
2     4.0
3     2.0
4     7.0
Name: ResultDir, dtype: float64

**AvgSpeed** - OK<BR />
I guess we are talking windspeed here

In [540]:
weather['AvgSpeed'].dtype

dtype('O')

The data type need to be changed into floats

In [541]:
weather['AvgSpeed'] = weather['AvgSpeed'].astype(float)
weather['AvgSpeed'].head()

0     9.2
1     9.6
2    13.4
3    13.4
4    11.9
Name: AvgSpeed, dtype: float64

In [542]:
np.sort(weather['AvgSpeed'].unique())

array([ 1.7,  1.9,  2. ,  2.1,  2.3,  2.4,  2.6,  2.7,  2.8,  2.9,  3. ,
        3.1,  3.2,  3.3,  3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,
        4.2,  4.3,  4.4,  4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,
        5.3,  5.4,  5.5,  5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,
        6.4,  6.5,  6.6,  6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,
        7.5,  7.6,  7.7,  7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,
        8.6,  8.7,  8.8,  8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,
        9.7,  9.8,  9.9, 10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7,
       10.8, 10.9, 11. , 11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8,
       11.9, 12. , 12.1, 12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9,
       13. , 13.1, 13.2, 13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. ,
       14.1, 14.2, 14.3, 14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15. , 15.1,
       15.2, 15.3, 15.4, 15.5, 15.6, 15.7, 15.8, 15.9, 16. , 16.1, 16.2,
       16.3, 16.4, 16.6, 16.7, 16.8, 16.9, 17.1, 17

No null, M or - (missing values)

**Turn all new columns from CodeSum into floats**

In [545]:
new_colums = ['','BCFG','BR','DZ','FG','FG+','FU','GR','HZ','MIFG','RA','SN','SQ','TS','TSRA','VCFG','VCTS']

for i in new_colums:
    weather[i] = weather[i].astype(float)

In [547]:
# rerun this cell when all the data types are correct and we know what columns we want to keep in the df.
weather.describe()

Station         Tmax         Tmin         Tavg       Depart  \
count  2920.000000  2920.000000  2920.000000  2920.000000  1464.000000   
mean      1.498630    76.131849    57.811644    67.223630     1.952186   
std       0.500084    11.477434    10.388355    10.555859     6.852047   
min       1.000000    41.000000    29.000000    36.000000   -17.000000   
25%       1.000000    69.000000    50.000000    60.000000    -3.000000   
50%       1.000000    78.000000    59.000000    69.000000     2.000000   
75%       2.000000    85.000000    66.000000    75.000000     7.000000   
max       2.000000   104.000000    83.000000    94.000000    23.000000   

          DewPoint      WetBulb         Heat         Cool     SnowFall  \
count  2920.000000  2920.000000  2920.000000  2920.000000  1464.000000   
mean     53.448288    59.313014     3.417466     5.641096     0.000888   
std      10.684779     9.289451     5.961108     6.109144     0.009385   
min      22.000000    32.000000     0.000000     0.000000     0.000000   
25%      46.000000    53.000000     0.000000     0.000000     0.000000   
50%      54.000000    61.000000     0.000000     4.000000     0.000000   
75%      62.000000    67.000000     5.000000    10.000000     0.000000   
max      75.000000    78.000000    29.000000    29.000000     0.100000   

          ...                GR           HZ         MIFG           RA  \
count     ...       2920.000000  2920.000000  2920.000000  2920.000000   
mean      ...          0.000342     0.089726     0.000685     0.317123   
std       ...          0.018506     0.285838     0.026167     0.465436   
min       ...          0.000000     0.000000     0.000000     0.000000   
25%       ...          0.000000     0.000000     0.000000     0.000000   
50%       ...          0.000000     0.000000     0.000000     0.000000   
75%       ...          0.000000     0.000000     0.000000     1.000000   
max       ...          1.000000     1.000000     1.000000     1.000000   

                SN           SQ           TS         TSRA         VCFG  \
count  2920.000000  2920.000000  2920.000000  2920.000000  2920.000000   
mean      0.002055     0.001027     0.034247     0.104110     0.001370   
std       0.045291     0.032042     0.181893     0.305455     0.036993   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              VCTS  
count  2920.000000  
mean      0.006507  
std       0.080416  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000  

[8 rows x 36 columns]

In [546]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2920 entries, 0 to 2943
Data columns (total 40 columns):
Station        2920 non-null float64
Date           2920 non-null datetime64[ns]
Tmax           2920 non-null float64
Tmin           2920 non-null float64
Tavg           2920 non-null float64
Depart         1464 non-null float64
DewPoint       2920 non-null float64
WetBulb        2920 non-null float64
Heat           2920 non-null float64
Cool           2920 non-null float64
Sunrise        2920 non-null object
Sunset         2920 non-null object
CodeSum        2920 non-null object
SnowFall       1464 non-null float64
PrecipTotal    2918 non-null float64
StnPressure    2920 non-null float64
SeaLevel       2920 non-null float64
ResultSpeed    2920 non-null float64
ResultDir      2920 non-null float64
AvgSpeed       2920 non-null float64
Year           2920 non-null int64
Month          2920 non-null int64
Day            2920 non-null int64
               2920 non-null float64
BCFG   

**Summary**

- The columns Depth and Water1 were removed from dataframe
- The Date column is datetime
- The columns Sunset and Sunrise are still strings and contain -. We will handle if we are going to use.
- We need to drop the CodeSum column later (kept it for now if something with the handmade dummies is  wrong), it's still strings
- Year, month, day are still int, do we want them to be floats? (really strange to have them as floats)

**Load the data to CSV**

In [548]:
weather.to_csv('weather_clean.csv')